Source/MQTTnet/Server/Internal/MqttClientSessionsManager.cs

```diff
@@ -544,6 +548,8 @@ MqttClient CreateClient(MqttConnectPacket connectPacket, IMqttChannelAdapter cha
                         await _eventContainer.ClientDisconnectedEvent.InvokeAsync(eventArgs).ConfigureAwait(false);
                     }
                 }
+
+                oldSession?.Dispose();
```

Comment:
During test, I found  if `SemaphoreSlim.Dispose()` is called here, `SemaphoreSlim.WaitAsync(ct)` failed to be cancelled and will never return

Reply:
Please do not change the dispose call yet. I am aware of this issue and already creating a new primitive as a replacement for the semaphore. After you reverted the Dispose call changes, we can merge this pull request because the other changes are very good and not dangerous.

Reply:
@chkr1011 Updated. 
Should `oldSession?.Dispose()` be kept as it is?
Seems the leak caused by `SemaphoreSlim.Dispose()` will only happen when debug in VS with `Task.s_currentActiveTasks` and will not happen in normal run.

Reply:
Please keep ist.